[PT] [ST] support non contiguous rank validation in sharded tensor

xunnanxu opened the pull request:
Summary:
Previously the validation logic assumes the sharded tensors' global ranks range from `[0 .. WS]`

This is true if we do 1d flat sharding.
But once we get into 2d+, the ranks may not be contiguous any more.

e.g.
```
[0, 2]
[1, 3]
```
The group size is 2 but ranks may be >= 2.

Going forward, the ST will be replaced by DTensor so it's less of an issue but this is just to make it work for stacks still relying on ST (like torchrec).

Test Plan:
added UT
CI

Differential Revision: D55671872




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

kwen2501 reviewed (approved):
Library code change looks good to me. 
Not sure if we'd want to add new tests given the deprecation state of Sharded Tensor.
But it may be okay if we plan to drop these tests all together at some point.

kwen2501 commented:
CI failure looks real

xunnanxu commented:
@pytorchbot merge

xunnanxu commented:
@pytorchbot merge

xunnanxu commented:
@pytorchbot rebase

xunnanxu commented:
@pytorchbot merge

xunnanxu committed:
[PT] [ST] support non contiguous rank validation in sharded tensor (#123230)

Summary:
Previously the validation logic assumes the sharded tensors' global ranks range from `[0 .. WS]`

This is true if we do 1d flat sharding.
But once we get into 2d+, the ranks may not be contiguous any more.

e.g.
```
[0, 2]
[1, 3]
```
The group size is 2 but ranks may be >= 2.

Going forward, the ST will be replaced by DTensor so it's less of an issue but this is just to make it work for stacks still relying on ST (like torchrec).


Test Plan:
added UT
CI





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

Reviewed By: kwen2501

Differential Revision: D55671872

Pulled By: xunnanxu

xunnanxu commented:
@pytorchbot merge
